substrate/frame/contracts/src/wasm/runtime.rs

```diff
@@ -2605,27 +2650,180 @@ pub mod env {
 	) -> Result<ReturnCode, TrapReason> {
-		use frame_support::dispatch::{extract_actual_weight, GetDispatchInfo};
+		use frame_support::dispatch::GetDispatchInfo;
 		ctx.charge_gas(RuntimeCosts::CopyFromContract(call_len))?;
 		let call: <E::T as Config>::RuntimeCall =
 			ctx.read_sandbox_memory_as_unbounded(memory, call_ptr, call_len)?;
-		let dispatch_info = call.get_dispatch_info();
-		let charged = ctx.charge_gas(RuntimeCosts::CallRuntime(dispatch_info.weight))?;
-		let result = ctx.ext.call_runtime(call);
-		let actual_weight = extract_actual_weight(&result, &dispatch_info);
-		ctx.adjust_gas(charged, RuntimeCosts::CallRuntime(actual_weight));
-		match result {
+		ctx.call_dispatchable::<CallRuntimeFailed, _>(call.get_dispatch_info(), |ext| {
+			ext.call_runtime(call)
+		})
+	}
+
+	/// Execute an XCM program locally, using the contract's address as the origin.
+	/// This is equivalent to dispatching `pallet_xcm::execute` through call_runtime, except that
+	/// the function is called directly instead of being dispatched.
+	///
+	/// # Parameters
+	///
+	/// - `msg_ptr`: the pointer into the linear memory where the message is placed.
+	/// - `msg_len`: the length of the message in bytes.
+	///
+	/// # Return Value
+	///
+	/// Returns `ReturnCode::Success` when the XCM was successfully executed. When the XCM
+	/// execution fails, `ReturnCode::CallRuntimeFailed` is returned.
+	#[unstable]
+	fn xcm_execute(
+		ctx: _,
+		memory: _,
+		msg_ptr: u32,
+		msg_len: u32,
+	) -> Result<ReturnCode, TrapReason> {
+		use crate::xcm::{CallOf, WeightInfo, Xcm};
+		use frame_support::dispatch::DispatchInfo;
+		use xcm::VersionedXcm;
+
+		ctx.charge_gas(RuntimeCosts::CopyFromContract(msg_len))?;
+		let message: VersionedXcm<CallOf<E::T>> =
+			ctx.read_sandbox_memory_as_unbounded(memory, msg_ptr, msg_len)?;
+
+		let execute_weight = <<E::T as Config>::Xcm as Xcm<E::T>>::WeightInfo::execute();
+		let weight = ctx.ext.gas_meter().gas_left().max(execute_weight);
+		let dispatch_info = DispatchInfo { weight, ..Default::default() };
+
+		ctx.call_dispatchable::<XcmExecutionFailed, _>(dispatch_info, |ext| {
+			<<E::T as Config>::Xcm as Xcm<E::T>>::execute(
+				ext.address(),
+				message,
+				weight.saturating_sub(execute_weight),
+			)
+		})
+	}
+
+	/// Send an XCM program from the contract to the specified destination.
+	/// This is equivalent to dispatching `pallet_xcm::send` through `call_runtime`, except that
+	/// the function is called directly instead of being dispatched.
+	///
+	/// # Parameters
+	///
+	/// - `dest_ptr`: the pointer into the linear memory where the [`xcm::VersionedMultiLocation`]
+	///   is placed.
+	/// - `call_ptr`: the pointer into the linear memory where the message is placed.
+	/// - `call_len`: the length of the message in bytes.
+	///
+	/// # Return Value
+	///
+	/// Returns `ReturnCode::Success` when the message was successfully sent. When the XCM
+	/// execution fails, `ReturnCode::CallRuntimeFailed` is returned.
```

Comment:
No other message is returned?

Reply:
Any data back we would want to communicate back, would  be provided through some kind of output pointer parameter that will point to a memory location where the returned value is written.
In that case since we merely call a dispatchable, there are no additional data to return.